Tools/build/bundle_ensurepip_wheels.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+
+"""
+Download wheels for 'ensurepip' packages from PyPI.
+
+When GitHub Actions executes the script, output is formatted accordingly.
+https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-notice-message
+"""
+
+import importlib.util
+import os
+import sys
+from hashlib import sha256
+from pathlib import Path
+from urllib.error import URLError
+from urllib.request import urlopen
+
+HOST = 'https://files.pythonhosted.org'
+ENSURE_PIP_ROOT = Path(__file__, "..", "..", "..", "Lib", "ensurepip").resolve()
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT = ENSURE_PIP_ROOT / "__init__.py"
+GITHUB_ACTIONS = "GITHUB_ACTIONS" in os.environ
+
+
+def print_notice(message: str) -> None:
+    if GITHUB_ACTIONS:
+        print(f"::notice::{message}", end="\n\n")
+    else:
+        print(message, file=sys.stderr)
+
+
+def print_error(message: str) -> None:
+    if GITHUB_ACTIONS:
+        print(f"::error::{message}", end="\n\n")
+    else:
+        print(message, file=sys.stderr)
+
+
+def download_wheels() -> int:
+    """Download wheels into bundle if they are not there yet."""
+
+    try:
+        projects = _get_projects()
+    except (AttributeError, TypeError):
```

Comment:
I think there's an abstraction leak here: it's hard to guess from looking at the `_get_projects()` function, what in it might trigger these exceptions. I'd recommend processing them inside that function and making obvious such places where the exceptions may occur. Instead, I'd convert both of the exceptions into something like an `ImportError` and handle just that. This would contribute to transparency of how that function works.